File: frontend/src/services/loyalty.service.ts
import { apiClient } from './api.service';
import type { LoyaltyAccount, LoyaltyTransaction, Reward, CreateRewardRequest } from '../types/loyalty.types';

export const loyaltyService = {
  async getMyAccount(): Promise<LoyaltyAccount> {
    return apiClient.get<LoyaltyAccount>('/loyalty/me');
  },

  async getMyHistory(): Promise<LoyaltyTransaction[]> {
    return apiClient.get<LoyaltyTransaction[]>('/loyalty/me/history');
  },

  async redeemReward(rewardId: number): Promise<LoyaltyAccount> {
    return apiClient.post<LoyaltyAccount>(`/loyalty/redeem/${rewardId}`);
  },

  async getAvailableRewards(): Promise<Reward[]> {
    return apiClient.get<Reward[]>('/loyalty/rewards');
  },

  async createReward(data: CreateRewardRequest): Promise<Reward> {
    return apiClient.post<Reward>('/loyalty/rewards', data);
  },

  async deleteReward(id: number): Promise<void> {
    return apiClient.delete<void>(`/loyalty/rewards/${id}`);
  },
};

File: backend/src/main/java/com/nischal/backend/service/LoyaltyService.java
package com.nischal.backend.service;

import com.nischal.backend.dto.loyalty.*;
import com.nischal.backend.entity.Order;

import java.util.List;

public interface LoyaltyService {

    /** Get or create loyalty account for the authenticated user */
    LoyaltyAccountResponse getMyAccount(String userEmail);

    /** Points history for the authenticated user */
    List<LoyaltyTransactionResponse> getMyHistory(String userEmail);

    /** Auto-called after order payment to award points */
    void earnPointsForOrder(Order order);

    /** Redeem a reward (deduct points) */
    LoyaltyAccountResponse redeemReward(String userEmail, Long rewardId);

    /** Award birthday bonus (called by scheduler) */
    void awardBirthdayBonus(Long userId);

    // Rewards catalogue
    List<RewardResponse> getAvailableRewards();
    RewardResponse createReward(CreateRewardRequest request);
    void deleteReward(Long rewardId);
}
